src/Microsoft.Diagnostics.Runtime/src/DataReaders/MacOS/MacOSProcessDataReader.cs

```diff
@@ -0,0 +1,364 @@
+
+            _task = task;
+
+            if (suspend)
+            {
+                status = Native.ptrace(Native.PT_ATTACH, processId);
+
+                if (status >= 0)
+                    status = Native.waitpid(processId, IntPtr.Zero, 0);
+
+                if (status < 0)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+                    throw new ClrDiagnosticsException($"Could not attach to process {processId}, errno: {errno}", errno);
+                }
+
+                _suspended = true;
+            }
+        }
+
+        ~MacOSProcessDataReader() => Dispose(false);
+
+        public string DisplayName => $"pid:{ProcessId:x}";
+
+        public bool IsThreadSafe => false;
+
+        public OSPlatform TargetPlatform => OSPlatform.OSX;
+
+        public Architecture Architecture => Architecture.Amd64;
+
+        public uint ProcessId { get; }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool _)
+        {
+            if (_disposed)
+                return;
+
+            if (_suspended)
+            {
+                int status = Native.ptrace(Native.PT_DETACH, (int)ProcessId);
+                if (status < 0)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+                    throw new ClrDiagnosticsException($"Could not detach from process {ProcessId}, errno: {errno}", errno);
+                }
+
+                _suspended = false;
+            }
+
+            _disposed = true;
+        }
+
+        public void FlushCachedData()
+        {
+            _memoryRegions = LoadMemoryRegions();
+        }
+
+        public IEnumerable<ModuleInfo> EnumerateModules()
+        {
+            int taskInfoCount = Native.TASK_DYLD_INFO_COUNT;
+            int kr = Native.task_info(_task, Native.TASK_DYLD_INFO, out Native.task_dyld_info dyldInfo, ref taskInfoCount);
+            if (kr != 0)
+                throw new ClrDiagnosticsException();
+
+            Native.dyld_all_image_infos infos = Read<Native.dyld_all_image_infos>(dyldInfo.all_image_info_addr);
+            for (uint i = 0; i < infos.infoArrayCount; i++)
+            {
+                Native.dyld_image_info info = Read<Native.dyld_image_info>(infos.infoArray, i);
+                ulong imageAddress = info.imageLoadAddress;
+                string imageFilePath = ReadNullTerminatedAscii(info.imageFilePath);
+                yield return new ModuleInfo(imageAddress, imageFilePath, true, 0, 0, ImmutableArray<byte>.Empty);
+            }
+
+            unsafe T Read<T>(ulong address, uint index = 0)
+                where T : unmanaged
+            {
+                T result;
+                if (Native.vm_read_overwrite(_task, address + index * (uint)sizeof(T), sizeof(T), &result, out _) != 0)
+                    throw new ClrDiagnosticsException();
```

Comment:
I don't think we should throw in Read, we should simply fail the read.  If we don't have permission to read from the process or in some way ALL calls to `vm_read_overwrite` fail, then we won't even get this far...we'll fail to create a dac instance.

```suggestion
                    return default;
```